src/quacc/recipes/vasp/core.py

```diff
@@ -4,18 +4,126 @@
+    bandgap: float = None,
+    vasprun: Vasprun = None,
+    **calc_kwargs,
+) -> VaspSchema:
+    """
+    Carry out a non-self-consistent field (NSCF) calculation.
+
+    Parameters
+    ----------
+    atoms
+        Atoms object.
+    preset
+        Preset to use from `quacc.calculators.vasp.presets`.
+    copy_files
+        Files to copy (and decompress) from source to the runtime directory.
+    kpoints_mode
+        Type of k-points mode. Options are "uniform" or "line". "boltztrap" may
+        be implemented in future.
+    calculate_optics
+        Whether to calculate optical properties.
+    bandgap
+        The band gap of the material.
+    **calc_kwargs
+        Custom kwargs for the Vasp calculator. Set a value to
+        `None` to remove a pre-existing key entirely. For a list of available
+        keys, refer to [quacc.calculators.vasp.vasp.Vasp][].
+
+    Returns
+    -------
+    VaspSchema
+        Dictionary of results from [quacc.schemas.vasp.vasp_summarize_run][].
+        See the type-hint for the data structure.
+    """
+
+    # Normalize kpoints_mode
+    kpoints_mode = kpoints_mode.lower()
+
+    # Validate kpoints_mode
+    supported_modes = ("line", "uniform")
+    if kpoints_mode not in supported_modes:
+        raise ValueError(f"Supported kpoints modes are: {', '.join(supported_modes)}")
+
+    calc_defaults = {
+        "lorbit": 11,
+        "lwave": False,
+        "lcharg": False,
+        "nsw": 0,
+        "isym": 0,
+        "icharg": 11,
+        "kspacing": None,
+        "nedos": 5001,
+    }
+
+    updates: dict[str, Any] = {}
+
+    if vasprun is not None:
+        nbands_factor = 1.2
+        nbands = int(np.ceil(vasprun.parameters["NBANDS"] * nbands_factor))
+        updates["nbands"] = nbands
+
+    def _get_nedos(vasprun: Vasprun | None, dedos: float) -> int:
+        """Automatic setting of nedos using the energy range and the energy step."""
+        if vasprun is None:
+            return 5001
+        emax = max(eigs.max() for eigs in vasprun.eigenvalues.values())
+        emin = min(eigs.min() for eigs in vasprun.eigenvalues.values())
+        return int((emax - emin) / dedos)
+
+    if kpoints_mode == "uniform":
+        # Automatic setting of NEDOS using the energy range and the energy step
+        n_edos = _get_nedos(vasprun, 0.005)  # Example value, adjust as needed
+        # Use tetrahedron method for DOS and optics calculations
+        updates.update({"ismear": -5, "isym": 2, "nedos": n_edos})
+    elif kpoints_mode == "line":
+        sigma = 0.2 if bandgap == 0 else 0.01
+        updates.update({"ismear": 0, "sigma": sigma})
+
+    if calculate_optics:
+        # LREAL not supported with LOPTICS = True; automatic NEDOS usually
+        # underestimates, so set it explicitly
+        n_edos = _get_nedos(vasprun, 0.0025)  # Example value, adjust as needed
+        updates.update(
+            {"loptics": True, "lreal": False, "cshift": 1e-5, "nedos": n_edos}
+        )
```

Comment:
I'd personally suggest we remove this block for now. If we want to allow optical calculations, we can have a dedicated `optics_job` that essentially calls the `nscf_job` with some updated parameters to add. This will help keep the `nscf_job` easier to read.